lib/dao.js

```diff
@@ -624,7 +624,197 @@ DataAccessObject.upsert = function(data, options, cb) {
+            }
+            callConnector();
+          }, update, options);
+        }
+        function callConnector() {
+          update = removeUndefined(update);
+          context = {
+            Model: Model,
+            where: ctx.where,
+            data: update,
+            currentInstance: inst,
+            hookState: ctx.hookState,
+            options: options,
+          };
+          Model.notifyObserversOf('persist', context, function(err) {
+            if (err) return done(err);
+            connector.upsertWithWhere(modelName, where, update, options, done);
+          });
+        }
+        function done(err, data, info) {
+          if (err) return cb(err);
+          var context = {
+            Model: Model,
+            data: data,
+            isNewInstance: info && info.isNewInstance,
+            hookState: ctx.hookState,
+            options: options,
+          };
+          Model.notifyObserversOf('loaded', context, function(err) {
+            if (err) return cb(err);
+            var obj;
+            if (data && !(data instanceof Model)) {
+              inst._initProperties(data, { persisted: true });
+              obj = inst;
+            } else {
+              obj = data;
+            }
+            if (err) {
+              cb(err, obj);
+            } else {
+              var context = {
+                Model: Model,
+                instance: obj,
+                isNewInstance: info ? info.isNewInstance : undefined,
+                hookState: hookState,
+                options: options,
+              };
+              Model.notifyObserversOf('after save', context, function(err) {
+                cb(err, obj);
+              });
+            }
+          });
+        }
+      });
+    } else {
+      var element = Object.keys(where)[0];
+      var filter = { 'where': where };
+      if (!Object.keys(data).length) {
+        var err = new Error('data object cannot be empty!');
+        err.statusCode = 400;
+        process.nextTick(function() { cb(err); });
+        return cb.promise;
+      }
 
+      if (!data.hasOwnProperty(element)) {
+        data[element] = where[element];
+      }
+      var pk = idName(Model);
+      self.find(filter, function(err, instances) {
+        if (err) return cb(err);
+        else {
+          var modelsLength = instances.length;
+          if (modelsLength === 0 && data.hasOwnProperty(pk)) {
+            self.create(data, options, cb);
+            return cb.promise;
+          } else if (modelsLength == 0 && data.hasOwnProperty(pk) == false) {
+            var error = new Error('Error message: No matching instance found.Found error while trying ' +
+              'to create a new instance.' +
+              'Primary key and value ' +
+              'is not provided in the payload');
+            cb(error);
+          } else if (modelsLength === 1) {
+            var pkName = idName(Model);
```

Comment:
It seems my previous comment disappeared and it has not been applied; `pkName` is not used.